linkerd/proxy/http/src/server.rs

```diff
@@ -145,120 +124,72 @@ where
-                let mut server = self.server.clone();
-                let drain = self.drain.clone();
-                Box::pin(async move {
-                    let (svc, closed) = SetClientHandle::new(io.peer_addr()?, http1);
-
-                    let mut conn = server
-                        .http1_only(true)
-                        .serve_connection(
-                            io,
-                            // Enable support for HTTP upgrades (CONNECT and websockets).
-                            upgrade::Service::new(svc, drain.clone()),
-                        )
-                        .with_upgrades();
-
-                    tokio::select! {
-                        res = &mut conn => {
-                            debug!(?res, "The client is shutting down the connection");
-                            res?
-                        }
-                        shutdown = drain.signal() => {
-                            debug!("The process is shutting down the connection");
-                            Pin::new(&mut conn).graceful_shutdown();
-                            shutdown.release_after(conn).await?;
-                        }
-                        () = closed => {
-                            debug!("The stack is tearing down the connection");
-                            Pin::new(&mut conn).graceful_shutdown();
-                            conn.await?;
+        match self.clone() {
+            Self::Http {
+                version,
+                service,
+                drain,
+                mut server,
+            } => Box::pin(async move {
+                debug!(?version, "Handling as HTTP");
+                let (svc, closed) = SetClientHandle::new(io.peer_addr()?, service);
+                match version {
+                    Version::Http1 => {
+                        // Enable support for HTTP upgrades (CONNECT and websockets).
+                        let mut conn = server
+                            .http1_only(true)
+                            .serve_connection(io, upgrade::Service::new(svc, drain.clone()))
+                            .with_upgrades();
+
+                        tokio::select! {
+                            res = &mut conn => {
+                                debug!(?res, "The client is shutting down the connection");
+                                res?
+                            }
+                            shutdown = drain.signal() => {
+                                debug!("The process is shutting down the connection");
+                                Pin::new(&mut conn).graceful_shutdown();
+                                shutdown.release_after(conn).await?;
+                            }
+                            () = closed => {
+                                debug!("The stack is tearing down the connection");
+                                Pin::new(&mut conn).graceful_shutdown();
+                                conn.await?;
+                            }
```

Comment:
it occurs to me that we probably _could_ stick the `select!` into an `async fn` that takes a drain, closed future, and connection future, so that this isn't repeated...but i dunno if that's worth it here, and it was like that before...

Reply:
let's do this separately -- the type constraints for this function would end up being pretty weighty. And this doesn't trigger the [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)).